### PR TITLE
`Nonterminal`-related cleanups

### PR DESCRIPTION
**Pull request by nnethercote**
In #114647 I am trying to remove `Nonterminal`. It has a number of preliminary cleanups that are worth merging even if #114647 doesn't merge, so let's do them in this PR.

r? @petrochenkov

### CODE REVIEW

File: compiler/rustc_ast/src/token.rs [hunk 1]
```diff
@@ -421,10 +421,11 @@ impl Token {
 
     /// Returns `true` if the token can appear at the start of an expression.
     pub fn can_begin_expr(&self) -> bool {
+        use Delimiter::*;
         match self.uninterpolate().kind {
             Ident(name, is_raw)              =>
                 ident_can_begin_expr(name, self.span, is_raw), // value name or keyword
-            OpenDelim(..)                     | // tuple, array or block
+            OpenDelim(Parenthesis | Brace | Bracket) | // tuple, array or block
```
[the remark this answers is not in the document]
Comment (nnethercote) [reply compiler/rustc_ast/src/token.rs:428]: It makes more sense in the context of #114647 where the invisibles case gets split into different kinds. I can just drop this commit from this PR.

[the remark this answers is not in the document]
Comment (nnethercote) [reply compiler/rustc_ast/src/token.rs:428]: Other than that, does this PR look ok?

[the remark this answers is not in the document]
Comment (petrochenkov) [reply compiler/rustc_ast/src/token.rs:428]: Yes, besides this comment r=me.

**Comment by nnethercote** [comment]
I removed the questionable commit.

@bors r=petrochenkov

**Pull request merged**